errors.go

```diff
@@ -0,0 +1,182 @@
+	ErrForbidden           = fmt.Errorf("%w", ErrUnauthorized)
+	ErrNotFound            = fmt.Errorf("%w: not_found", ErrResponse)
+	ErrNotAcceptable       = fmt.Errorf("%w: not_acceptable", ErrResponse)
+	ErrConflict            = fmt.Errorf("%w: conflict", ErrResponse)
+	ErrUnprocessableEntity = fmt.Errorf("%w: unprocessable_entity", ErrResponse)
+	ErrTooManyRequests     = fmt.Errorf("%w: too_many_requests", ErrResponse)
+
+	ErrInternalServerError = fmt.Errorf(
+		"%w: internal_server_error", ErrResponse,
+	)
+	ErrBadGateway         = fmt.Errorf("%w: bad_gateway", ErrResponse)
+	ErrServiceUnavailable = fmt.Errorf("%w: service_unavailable", ErrResponse)
+	ErrGatewayTimeout     = fmt.Errorf("%w: gateway_timeout", ErrResponse)
+)
+
+type responseErrorBody struct {
+	Error *ResponseError `json:"error,omitempty"`
+}
+
+type ResponseError struct {
+	parent error
+
+	Code        string          `json:"code,omitempty"`
+	Description string          `json:"description,omitempty"`
+	Detail      json.RawMessage `json:"detail,omitempty"`
+}
+
+func NewResponseError(
+	httpStatus int,
+	code string,
+	description string,
+	rawDetail json.RawMessage,
+) *ResponseError {
+	var parent error
+	switch httpStatus {
+	case http.StatusBadRequest:
+		parent = ErrBadRequest
+	case http.StatusUnauthorized:
+		parent = ErrUnauthorized
+	case http.StatusForbidden:
+		parent = ErrForbidden
+	case http.StatusNotFound:
+		parent = ErrResourceNotFound
+	case http.StatusNotAcceptable:
+		parent = ErrNotAcceptable
+	case http.StatusConflict:
+		parent = ErrConflict
+	case http.StatusUnprocessableEntity:
+		parent = ErrUnprocessableEntity
+	case http.StatusTooManyRequests:
+		parent = ErrTooManyRequests
+	case http.StatusInternalServerError:
+		parent = ErrInternalServerError
+	case http.StatusBadGateway:
+		parent = ErrBadGateway
+	case http.StatusServiceUnavailable:
+		parent = ErrServiceUnavailable
+	case http.StatusGatewayTimeout:
+		parent = ErrGatewayTimeout
+	default:
+		parent = ErrUnknown
+	}
+
+	return &ResponseError{
+		parent:      parent,
+		Code:        code,
+		Description: description,
+		Detail:      rawDetail,
+	}
+}
+
+func (s *ResponseError) Error() string {
+	out := s.Code
+	if s.Description != "" {
+		out += ": " + s.Description
+	}
+
+	// When RawDetail is not a empty JSON object ("{}" ), we prettify and
+	// include it in the error.
+	if len(s.Detail) > 2 {
+		buf := &bytes.Buffer{}
+		_ = json.Indent(buf, s.Detail, "", "  ")
+		out += " -- " + buf.String()
+	}
+
+	return out
+}
+
+func (s *ResponseError) Is(target error) bool {
+	return errors.Is(s.parent, target)
+}
+
+func (s *ResponseError) Unwrap() error {
+	return s.parent
+}
+
+// CommonError handles common logic shared between all API-based error types.
+type CommonError struct {
+	parent error
+
+	Code        string `json:"code,omitempty"`
+	Description string `json:"description,omitempty"`
+}
+
+func NewCommonError(parent error, code, description string) *CommonError {
```

Comment:
Perhaps we ought to return a value type here? It seems weird that we have to use indirection in the generated code.

Reply:
Yeah, agreed. I was kind of undecided about it cause it felt wasteful copying the whole struct, but I think only the `Code` and `Description` fields which would be copied as the `parent` field should be a pointer to a `error`, it's probably fine.